defi-pulse/src/symbols/index.ts

```diff
@@ -0,0 +1,80 @@
+import fs from 'fs'
+import { join } from 'path'
+import { ethers, utils } from 'ethers'
+import { util } from '@chainlink/ea-bootstrap'
+
+const directoryPath = './src/symbols/directory.json'
+
+type Directory = Record<string, string>
+
+const getDirectory = (): Directory => {
+  const absolutePath = join(process.cwd(), directoryPath)
+  const buffer = fs.readFileSync(absolutePath, 'utf8')
+  return JSON.parse(buffer.toString())
+}
+
+const updateDirectory = (address: string, symbol: string): void => {
+  const absolutePath = join(process.cwd(), directoryPath)
+  const directory = cachedDirectory().directory
+  directory[address] = symbol
+  fs.writeFileSync(absolutePath, JSON.stringify(directory))
+  cachedDirectory().update()
+}
+
+const ERC20ABI = ['function symbol() view returns (string)']
+const ERC20ABI2 = [
+  {
+    constant: true,
+    inputs: [],
+    name: 'symbol',
+    outputs: [
+      {
+        name: '',
+        type: 'bytes32',
+      },
+    ],
+    payable: false,
+    stateMutability: 'view',
+    type: 'function',
+  },
+]
+
+// Would be better from etherscan
+const getRpcTokenSymbol = async (address: string): Promise<string> => {
+  console.log('calling blockchain for symbol')
+  const rpcUrl = util.getRequiredEnv('RPC_URL')
+  const provider = new ethers.providers.JsonRpcProvider(rpcUrl)
+  try {
+    const erc20 = new ethers.Contract(address, ERC20ABI, provider)
+    const symbol = await erc20.symbol()
+    return symbol
+  } catch (e) {
+    const erc20 = new ethers.Contract(address, ERC20ABI2, provider)
+    const symbol = await erc20.symbol()
+    return utils.parseBytes32String(symbol)
+  }
+}
+
+function memoizeDirectory() {
+  let directory = getDirectory()
+  return () => {
+    return { directory, update: (): Directory => (directory = getDirectory()) }
+  }
+}
+
+const cachedDirectory = memoizeDirectory()
```

Comment:
There are a few issues here:

- The caching implementation should not be specific to one use case or to one adapter.
- Most of these adapters are run in a serverless environment, where the filesystem is very volatile and short-living as are the serverless functions themselves. This means that you can cache work during the request, but not cross requests as the env will probably be torn down by the underlying execution scheduler.
- We already have 2x cache implementation that sits in front of all adapters (local in-memory & Redis). They cache adapter `input => output`. If we would need additional inter-request caching to speed up some work (like this example), then we should write some code to expose that caching service as an API to the adapter for use. This would be the right path forward, but it's not easy and requires some refactoring.
- The implementation should probably take into consideration multiple networks (testnet vs mainnet vs...)

My suggestion would be to KISS here. It's ok to ship this adapter with a DB of symbols, but I would use it for reading not for writing to it because of the ephemeral environment issue. Just read it to the local in-memory cache when the adapter starts, and use that as a lookup table for symbols. You can also write to that local in-memory cache if it makes sense to avoid duplicate calls to Ethereum nodes during one request but it probably does not make sense (because a specific Index has one asset repeating only once) so just read.